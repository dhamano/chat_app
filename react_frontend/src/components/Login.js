import React, { useState } from 'react';
import { NavLink } from 'react-router-dom';
import { login, register } from '../services';
import { setLocalStorage } from '../utilities';

const Login = props => {
    const [password, setPassword] = useState(false);
    const [confirmPassword, setConfirmPassword] = useState(false);
    const [passwordErr, setPasswordErr] = useState(false);

    function handleOnChange (e) {
        if(e.target.name === 'username') props.loginRegVals.setUsername(e.target.value);
        if(e.target.name === 'password') setPassword(e.target.value);
        if(e.target.name === 'confirm-password') setConfirmPassword(e.target.value);
    };

    async function handleSubmitLogin (e) {
        e.preventDefault();
        console.log('login', props.loginRegVals.username, password);
        await login({ username: props.loginRegVals.username, password })
                            .then(res => {
                                console.log('login res', res);
                                setLocalStorage("token", res.data.token);
                                setLocalStorage("username", res.data.username);
                                props.history.push('/')
                            })
    }

    async function handleSubmitRegister (e) {
        e.preventDefault();
        if( password !== confirmPassword) {
            setPasswordErr(true);
            return;
        }
        await register({ username: props.loginRegVals.username, password })
                            .then(res => {
                                console.log('register', res);
                                if ( res.status === 201 ) {
                                    setLocalStorage("token", res.data.token);
                                    setLocalStorage("username", res.data.username);
                                    props.history.push('/')
                                } else {
                                    console.log('ERROR');
                                }
                            })
    }

    return (
        <div className="login-reg-page">
            <div className="login-reg-container">
                <div className={`form-container ${props.loginRegVals.loginOrReg === 'Login' ? 'login' : 'register'}`}>
                    <ul>
                        <li><NavLink onClick={props.loginRegVals.onClickHandler} exact to="/">Login</NavLink></li>
                        <li><NavLink onClick={props.loginRegVals.onClickHandler} to="/register">Register</NavLink></li>
                    </ul>
                    { props.loginRegVals.loginOrReg === "Login" ? (
                    <div className="login">
                        <form onSubmit={handleSubmitLogin}>
                            <div>
                                <input id="username" onChange={handleOnChange} value={!props.loginRegVals.username ? '' : props.loginRegVals.username} name="username" type="text" placeholder="username" autoComplete="username" required />
                                <label htmlFor="username">username</label>
                            </div>
                            <div>
                                <input id="password" onChange={handleOnChange} value={!password ? '' : password} name="password" type="password" placeholder="password" autoComplete="current-password" required />
                                <label htmlFor="password">password</label>
                            </div>
                            <button type="submit">Login</button>
                        </form>
                    </div>
                    ) : (
                    <div className="register">
                        <form onSubmit={handleSubmitRegister}>
                            <div>
                                <input id="reg-username" onChange={handleOnChange} value={!props.loginRegVals.username ? '' : props.loginRegVals.username} name="username" type="text" placeholder="username" autoComplete="username" required />
                                <label htmlFor="reg-username">username</label>
                            </div>
                            <div>
                                <input id="reg-password" className={ passwordErr ? 'error' : '' } onChange={handleOnChange} value={!password ? '' : password} name="password" type="password" placeholder="password" autoComplete="new-password" required />
                                <label htmlFor="reg-password">password</label>
                            </div>
                            <div>
                                <input id="confirmPassword" className={ passwordErr ? 'error' : '' } onChange={handleOnChange} value={!confirmPassword ? '' : confirmPassword} name="confirm-password" type="password" placeholder="confirm password" autoComplete="new-password" required />
                                <label htmlFor="confirmPassword">confirm password</label>
                            </div>
                            <button type="submit">Register</button>
                        </form>
                    </div>
                    )}
                </div>
            </div>
        </div>
    )
};

export default Login;