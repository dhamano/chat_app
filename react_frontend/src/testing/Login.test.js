import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import toJSON from 'enzyme-to-json';
import { shallow, mount, render } from 'enzyme';

import Login from '../components/Login';

describe('<Login /> at Register', () => {
    let loginProps = {
            loginRegVals : {
                loginOrReg: 'Register'
            }
        }

        it('should have only one form', () => {
            const wrapper = shallow(<Login {...loginProps} />);
    
            expect(wrapper.find("form").exists()).toBe(true);
            expect(wrapper.find("form")).toHaveLength(1);
        });
    
        it('should have three input fields', () => {
            const wrapper = shallow(<Login {...loginProps} />);
    
            expect(wrapper.find("input").exists()).toBe(true);
            expect(wrapper.find("input")).toHaveLength(3);
        });
    
        it('should contain a submit button', () => {
            const wrapper = shallow(<Login {...loginProps} />);
    
            expect(wrapper.find("button").exists()).toBe(true);
            expect(wrapper.find("button")).toHaveLength(1);
        });
    
        it('should contain a username field and it should be empty', () => {
            const wrapper = shallow(<Login {...loginProps} />);
    
            expect(wrapper.find("#reg-username").exists()).toBe(true);
            expect(wrapper.find("#reg-username").props().value).toEqual('');
        });
    
        it('should contain a password field and it should be empty', () => {
            const wrapper = shallow(<Login {...loginProps} />);
    
            expect(wrapper.find("#reg-password").exists()).toBe(true);
            expect(wrapper.find("#reg-password").props().value).toEqual('');
        });

});

describe('<Login /> at login', () => {
    let loginProps = {
            loginRegVals : {
                loginOrReg: 'Login'
            }
        }

    it('renders', () => {
        const wrapper = shallow(<Login {...loginProps} />);
        expect(wrapper.exists()).toBe(true);
    });

    it('should match snapshot since login is deault', () => {
        const wrapper = shallow(<Login { ...loginProps } />);
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('should have only one form', () => {
        const wrapper = shallow(<Login {...loginProps} />);

        expect(wrapper.find("form").exists()).toBe(true);
        expect(wrapper.find("form")).toHaveLength(1);
    });

    it('should have two input fields', () => {
        const wrapper = shallow(<Login {...loginProps} />);

        expect(wrapper.find("input").exists()).toBe(true);
        expect(wrapper.find("input")).toHaveLength(2);
    });

    it('should contain a submit button', () => {
        const wrapper = shallow(<Login {...loginProps} />);

        expect(wrapper.find("button").exists()).toBe(true);
        expect(wrapper.find("button")).toHaveLength(1);
    });

    it('should contain a username field and it should be empty', () => {
        const wrapper = shallow(<Login {...loginProps} />);

        expect(wrapper.find("#username").exists()).toBe(true);
        expect(wrapper.find("#username").props().value).toEqual('');
    });

    it('should contain a password field and it should be empty', () => {
        const wrapper = shallow(<Login {...loginProps} />);

        expect(wrapper.find("#password").exists()).toBe(true);
        expect(wrapper.find("#password").props().value).toEqual('');
    });
});
