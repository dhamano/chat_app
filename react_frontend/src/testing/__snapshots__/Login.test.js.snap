// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> at login should match snapshot since login is deault 1`] = `
<div
  className="login-reg-page"
>
  <div
    className="login-reg-container"
  >
    <div
      className="form-container login"
    >
      <ul>
        <li>
          <NavLink
            exact={true}
            to="/"
          >
            Login
          </NavLink>
        </li>
        <li>
          <NavLink
            to="/register"
          >
            Register
          </NavLink>
        </li>
      </ul>
      <div
        className="login"
      >
        <form
          onSubmit={[Function]}
        >
          <div>
            <input
              autoComplete="username"
              id="username"
              name="username"
              onChange={[Function]}
              placeholder="username"
              required={true}
              type="text"
              value=""
            />
            <label
              htmlFor="username"
            >
              username
            </label>
          </div>
          <div>
            <input
              autoComplete="current-password"
              id="password"
              name="password"
              onChange={[Function]}
              placeholder="password"
              required={true}
              type="password"
              value=""
            />
            <label
              htmlFor="password"
            >
              password
            </label>
          </div>
          <button
            type="submit"
          >
            Login
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
